### PR TITLE
[FEATURE] Range awareness for alignment algorithm

### PR DESCRIPTION
**Pull request by svnbgnk**
Issue #1328

Add new call operator to alignment_algorithm

-that takes a range, calls the two sequences call operator on each sequence pair and writes the results -into a vector, return the vector
-no tests needed, add new documentation

### CODE REVIEW

File: include/seqan3/alignment/pairwise/alignment_algorithm.hpp [hunk 1]
```diff
@@ -190,6 +191,53 @@ class alignment_algorithm :
...
+     *
+     * \param[in] job_range   The range of underlying sequence pairs to be aligned.
+     *
+     * \returns A seqan3::alignment_result with the requested alignment outcomes.
+     *
+     * \details
+     *
+     * The algorithm always computes a pairwise alignment of two sequences over either a regular alphabet or
+     * packed alphabets in a SIMD vector. In the latter case an inter-vectorisation layout
+     * is used to compute l many pairwise alignments in parallel using special extended register instructions.
```
Comment (joergi-w) [inline include/seqan3/alignment/pairwise/alignment_algorithm.hpp]: What is `l`?

File: include/seqan3/alignment/pairwise/alignment_algorithm.hpp [hunk 2]
```diff
@@ -190,6 +191,53 @@ class alignment_algorithm :
         return alignment_result{make_result(idx, first_range, second_range, cache)};
     }
 
+    /*!\brief Invokes the actual alignment computation given two sequences.
+     * \tparam    job_range_type  The type of underlying sequence pairs; must model std::ranges::view.
```
Comment (rrahn) [inline include/seqan3/alignment/pairwise/alignment_algorithm.hpp]:
```suggestion
     * \tparam    indexed_sequence_pairs_t  The type of the range over indexed sequence pairs; must model std::ranges::forward_range.
```

File: include/seqan3/alignment/pairwise/alignment_algorithm.hpp [hunk 3]
```diff
@@ -190,6 +191,53 @@ class alignment_algorithm :
...
+     *
+     * \param[in] job_range   The range of underlying sequence pairs to be aligned.
```
Comment (rrahn) [inline include/seqan3/alignment/pairwise/alignment_algorithm.hpp]:
```suggestion
     * \param[in] indexed_sequence_pairs  The range of indexed sequence pairs to align.
```

File: include/seqan3/alignment/pairwise/alignment_algorithm.hpp [hunk 4]
```diff
@@ -190,6 +191,53 @@ class alignment_algorithm :
...
+     *
+     * \returns A seqan3::alignment_result with the requested alignment outcomes.
```
Comment (rrahn) [inline include/seqan3/alignment/pairwise/alignment_algorithm.hpp]:
```suggestion
     * \returns A std::vector over seqan3::alignment_result containing the results of the computed alignments.
```

File: include/seqan3/alignment/pairwise/alignment_algorithm.hpp [hunk 5]
```diff
@@ -190,6 +191,53 @@ class alignment_algorithm :
...
+     * and at most \f$ O(N^2) \f$ space.
+     */
+
+    template <std::ranges::forward_range job_range_type>
+    auto operator()(job_range_type && job_range)
+    {
+        static_assert(std::ranges::view<job_range_type>, "Expected a view!");
+
+         using fst_range_t = decltype(get<0>(*std::ranges::begin(job_range).first));
+         using snd_range_t = decltype(get<1>(*std::ranges::begin(job_range).first));
```
[the remark this answers is not in the document]
Comment (rrahn) [reply include/seqan3/alignment/pairwise/alignment_algorithm.hpp]: Please call it sequence1_t and sequence2_t.

File: include/seqan3/alignment/pairwise/alignment_algorithm.hpp [hunk 6]
```diff
@@ -190,6 +191,53 @@ class alignment_algorithm :
...
+     * The code always runs in \f$ O(N^2) \f$ time and depending on the configuration requires at least \f$ O(N) \f$
+     * and at most \f$ O(N^2) \f$ space.
+     */
+
+    template <std::ranges::forward_range job_range_type>
+    auto operator()(job_range_type && job_range)
+    {
+        static_assert(std::ranges::view<job_range_type>, "Expected a view!");
+
+         using fst_range_t = decltype(get<0>(*std::ranges::begin(job_range).first));
```
Comment (rrahn) [inline include/seqan3/alignment/pairwise/alignment_algorithm.hpp]:
```suggestion
         using sequence_pairs_t = std::tuple_element_t<0, std::ranges::value_t<indexed_sequence_pairs_t>>;
         using sequence1_t = std::tuple_element_t<0, sequence_pairs_t>;
         using sequence2_t = std::tuple_element_t<1, sequence_pairs_t>;
```

File: include/seqan3/alignment/pairwise/alignment_algorithm.hpp (hunk 1 above)
Comment (rrahn) [reply include/seqan3/alignment/pairwise/alignment_algorithm.hpp]: @joergi-w this is ok. I have the clean up PR open. Depending on what will merged before I will then fix the documentation accordingly. It is also not accurate anymore.

**Comment by svnbgnk** [comment]
I change quite a bit to make it more similar to the range aware operator of edit_distance_algorithm. I really did not like that they had different Input parameters.

**Comment by svnbgnk** [comment]
Oh there seems to be a problem with std::ranges::value_t.

File: include/seqan3/alignment/pairwise/alignment_algorithm.hpp [hunk 7]
```diff
@@ -190,6 +191,56 @@ class alignment_algorithm :
         return alignment_result{make_result(idx, first_range, second_range, cache)};
     }
 
+    /*!\brief Invokes the actual alignment computation given two sequences.
+     * \tparam    sequence_pairs_t  The type of the range over indexed sequence pairs; must model std::ranges::forward_range.
+     *  and the value type of the range must model seqan3::tuple_like and with a tuple size of 2.
+     *
+     *
```
Comment (joergi-w) [inline include/seqan3/alignment/pairwise/alignment_algorithm.hpp]: Please remove consecutive empty lines.

File: include/seqan3/alignment/pairwise/alignment_algorithm.hpp [hunk 8]
```diff
@@ -190,6 +191,56 @@ class alignment_algorithm :
         return alignment_result{make_result(idx, first_range, second_range, cache)};
     }
 
+    /*!\brief Invokes the actual alignment computation given two sequences.
+     * \tparam    sequence_pairs_t  The type of the range over indexed sequence pairs; must model std::ranges::forward_range.
+     *  and the value type of the range must model seqan3::tuple_like and with a tuple size of 2.
```
Comment (joergi-w) [inline include/seqan3/alignment/pairwise/alignment_algorithm.hpp]: You continue the sentence after a full stop.

Comment (joergi-w) [inline include/seqan3/alignment/pairwise/alignment_algorithm.hpp]:
```suggestion
     *  and the value type of the range must model seqan3::tuple_like and have tuple size 2.
```

Comment (joergi-w) [reply include/seqan3/alignment/pairwise/alignment_algorithm.hpp]: or better: `and std::tuple_size must evaluate to 2.`

**Comment by rrahn** [comment]
superseded by #1356

**Pull request closed without merging**